File: ts-api/src/anlas.ts
/**
 * NovelAI Anlas コスト計算モジュール
 * 純粋な計算ロジックのみ。API呼び出し・非同期処理・外部依存なし。
 */

import {
  OPUS_FREE_PIXELS,
  OPUS_FREE_MAX_STEPS,
  OPUS_MIN_TIER,
  MAX_COST_PER_IMAGE,
  MIN_COST_PER_IMAGE,
  GRID_SIZE,
  VIBE_BATCH_PRICE,
  VIBE_FREE_THRESHOLD,
  VIBE_ENCODE_PRICE,
  CHAR_REF_PRICE,
  INPAINT_THRESHOLD_RATIO,
  V4_COST_COEFF_LINEAR,
  V4_COST_COEFF_STEP,
  AUGMENT_FIXED_STEPS,
  AUGMENT_MIN_PIXELS,
  BG_REMOVAL_MULTIPLIER,
  BG_REMOVAL_ADDEND,
  UPSCALE_COST_TABLE,
  UPSCALE_OPUS_FREE_PIXELS,
  MAX_PIXELS,
} from './constants';


// =============================================================================
// 型定義
// =============================================================================

/** サブスクリプションティア（0=Free, 1=Tablet, 2=Scroll, 3=Opus） */
export type SubscriptionTier = 0 | 1 | 2 | 3;

/** SMEAモード */
export type SmeaMode = 'off' | 'smea' | 'smea_dyn';

/** 生成モード */
export type GenerationMode = 'txt2img' | 'img2img' | 'inpaint';

/** Augmentツールタイプ */
export type AugmentToolType = 'colorize' | 'declutter' | 'emotion' | 'sketch' | 'lineart' | 'bg-removal';

/** 画像生成コスト計算パラメータ */
export type GenerationCostParams = {
  width: number;
  height: number;
  steps: number;
  smea?: SmeaMode;              // デフォルト: 'off'
  mode?: GenerationMode;        // デフォルト: 'txt2img'
  strength?: number;            // デフォルト: 1.0
  nSamples?: number;            // デフォルト: 1
  tier?: SubscriptionTier;      // デフォルト: 0
  charRefCount?: number;        // デフォルト: 0
  vibeCount?: number;           // デフォルト: 0
  vibeUnencodedCount?: number;  // デフォルト: 0
  maskWidth?: number;           // Inpaintマスク寸法
  maskHeight?: number;
};

/** 画像生成コスト計算結果（内訳付き） */
export type GenerationCostResult = {
  baseCost: number;
  smeaMultiplier: number;
  perImageCost: number;
  strengthMultiplier: number;
  adjustedCost: number;
  isOpusFree: boolean;
  billableImages: number;
  generationCost: number;
  charRefCost: number;
  vibeEncodeCost: number;
  vibeBatchCost: number;
  totalCost: number;
  error: boolean;
  errorCode: number | null;
};

/** Augmentコスト計算パラメータ */
export type AugmentCostParams = {
  tool: AugmentToolType;
  width: number;
  height: number;
  tier?: SubscriptionTier;      // デフォルト: 0
};

/** Augmentコスト計算結果 */
export type AugmentCostResult = {
  originalPixels: number;
  adjustedWidth: number;
  adjustedHeight: number;
  adjustedPixels: number;
  baseCost: number;
  finalCost: number;
  isOpusFree: boolean;
  effectiveCost: number;
};

/** アップスケールコスト計算パラメータ */
export type UpscaleCostParams = {
  width: number;
  height: number;
  tier?: SubscriptionTier;      // デフォルト: 0
};

/** アップスケールコスト計算結果 */
export type UpscaleCostResult = {
  pixels: number;
  cost: number | null;
  isOpusFree: boolean;
  error: boolean;
  errorCode: number | null;
};

/** Inpaintサイズ補正結果 */
export type InpaintCorrectionResult = {
  corrected: boolean;
  width: number;
  height: number;
};


// =============================================================================
// 基本計算関数
// =============================================================================

/**
 * V4モデルの基本コストを計算
 * ピクセル数とステップ数に基づく線形コスト計算
 * @param width 画像幅
 * @param height 画像高さ
 * @param steps ステップ数
 * @returns 基本コスト（Anlas）
 */
export function calcV4BaseCost(width: number, height: number, steps: number): number {
  const pixels = width * height;
  return Math.ceil(V4_COST_COEFF_LINEAR * pixels + V4_COST_COEFF_STEP * pixels * steps);
}

/**
 * SMEAモードに対応するコスト乗数を返す
 * @param mode SMEAモード
 * @returns 乗数（1.0 / 1.2 / 1.4）
 */
export function getSmeaMultiplier(mode: SmeaMode): number {
  switch (mode) {
    case 'smea_dyn': return 1.4;
    case 'smea':     return 1.2;
    case 'off':      return 1.0;
  }
}

/**
 * Opus無料生成の条件を判定
 * @param width 画像幅（元のリクエストサイズ）
 * @param height 画像高さ（元のリクエストサイズ）
 * @param steps ステップ数
 * @param charRefCount キャラクター参照数
 * @param tier サブスクリプションティア
 * @returns Opus無料かどうか
 */
export function isOpusFreeGeneration(
  width: number,
  height: number,
  steps: number,
  charRefCount: number,
  tier: SubscriptionTier,
): boolean {
  return charRefCount === 0
    && width * height <= OPUS_FREE_PIXELS
    && steps <= OPUS_FREE_MAX_STEPS
    && tier >= OPUS_MIN_TIER;
}

/**
 * Vibeバッチコストを計算
 * 無料枠（VIBE_FREE_THRESHOLD）を超えた分のみ課金
 * @param enabledVibeCount 有効なVibe数
 * @returns Vibeバッチコスト（Anlas）
 */
export function calcVibeBatchCost(enabledVibeCount: number): number {
  return Math.max(0, enabledVibeCount - VIBE_FREE_THRESHOLD) * VIBE_BATCH_PRICE;
}

/**
 * キャラクター参照コストを計算
 * @param charRefCount キャラクター参照数
 * @param nSamples 生成枚数
 * @returns キャラクター参照コスト（Anlas）
 */
export function calcCharRefCost(charRefCount: number, nSamples: number): number {
  return CHAR_REF_PRICE * charRefCount * nSamples;
}


// =============================================================================
// ピクセル数調整関数
// =============================================================================

/**
 * 最小ピクセル数まで拡大（アスペクト比維持）
 * グリッドスナップなし、Math.floorのみ適用
 * @param width 元の幅
 * @param height 元の高さ
 * @param minPixels 最小ピクセル数
 * @returns 調整後の幅・高さ・ピクセル数
 */
export function expandToMinPixels(
  width: number,
  height: number,
  minPixels: number,
): { width: number; height: number; pixels: number } {
  const pixels = width * height;
  if (pixels >= minPixels) {
    return { width, height, pixels };
  }
  const scale = Math.sqrt(minPixels / pixels);
  const newW = Math.floor(width * scale);
  const newH = Math.floor(height * scale);
  return { width: newW, height: newH, pixels: newW * newH };
}

/**
 * 最大ピクセル数まで縮小（アスペクト比維持）
 * グリッドスナップなし、Math.floorのみ適用
 * @param width 元の幅
 * @param height 元の高さ
 * @param maxPixels 最大ピクセル数
 * @returns 調整後の幅・高さ・ピクセル数
 */
export function clampToMaxPixels(
  width: number,
  height: number,
  maxPixels: number,
): { width: number; height: number; pixels: number } {
  const pixels = width * height;
  if (pixels <= maxPixels) {
    return { width, height, pixels };
  }
  const scale = Math.sqrt(maxPixels / pixels);
  const newW = Math.floor(width * scale);
  const newH = Math.floor(height * scale);
  return { width: newW, height: newH, pixels: newW * newH };
}


// =============================================================================
// Inpaint補正
// =============================================================================

/**
 * Inpaintマスクサイズの補正を計算
 * マスクが閾値より小さい場合、OPUS_FREE_PIXELSまで拡大してグリッドスナップ
 * @param maskWidth マスク幅
 * @param maskHeight マスク高さ
 * @returns 補正結果（補正されたかどうか・幅・高さ）
 */
export function calcInpaintSizeCorrection(
  maskWidth: number,
  maskHeight: number,
): InpaintCorrectionResult {
  const pixels = maskWidth * maskHeight;
  const threshold = OPUS_FREE_PIXELS * INPAINT_THRESHOLD_RATIO;

  if (pixels >= threshold) {
    return { corrected: false, width: maskWidth, height: maskHeight };
  }

  const scale = Math.sqrt(OPUS_FREE_PIXELS / pixels);
  const newW = Math.floor(Math.floor(maskWidth * scale) / GRID_SIZE) * GRID_SIZE;
  const newH = Math.floor(Math.floor(maskHeight * scale) / GRID_SIZE) * GRID_SIZE;

  return { corrected: true, width: newW, height: newH };
}


// =============================================================================
// メイン：画像生成コスト計算
// =============================================================================

/**
 * 画像生成のAnlasコストを計算（メインオーケストレータ）
 * すべてのコスト要素（基本コスト、SMEA、strength、Opus無料、Vibe、キャラクター参照）を
 * 統合して最終コストと内訳を返す。
 * @param params 生成パラメータ
 * @returns コスト内訳を含む計算結果
 */
export function calculateGenerationCost(params: GenerationCostParams): GenerationCostResult {
  // デフォルト値の適用
  const smea = params.smea ?? 'off';
  const mode = params.mode ?? 'txt2img';
  const strength = params.strength ?? 1.0;
  const nSamples = params.nSamples ?? 1;
  const tier = params.tier ?? 0;
  const charRefCount = params.charRefCount ?? 0;
  const vibeCount = params.vibeCount ?? 0;
  const vibeUnencodedCount = params.vibeUnencodedCount ?? 0;

  // 有効な幅・高さの決定（Inpaintマスク補正）
  let effectiveWidth = params.width;
  let effectiveHeight = params.height;

  if (mode === 'inpaint' && params.maskWidth != null && params.maskHeight != null) {
    const correction = calcInpaintSizeCorrection(params.maskWidth, params.maskHeight);
    if (correction.corrected) {
      effectiveWidth = correction.width;
      effectiveHeight = correction.height;
    }
  }

  // 基本コスト計算
  const baseCost = calcV4BaseCost(effectiveWidth, effectiveHeight, params.steps);

  // SMEA乗数
  const smeaMultiplier = getSmeaMultiplier(smea);
  const perImageCost = baseCost * smeaMultiplier;

  // strength乗数（txt2imgは常に1.0）
  let strengthMultiplier: number;
  switch (mode) {
    case 'txt2img':
      strengthMultiplier = 1.0;
      break;
    case 'img2img':
    case 'inpaint':
      strengthMultiplier = strength;
      break;
  }

  // 調整後コスト（最低MIN_COST_PER_IMAGE保証）
  const adjustedCost = Math.max(Math.ceil(perImageCost * strengthMultiplier), MIN_COST_PER_IMAGE);

  // エラー判定（最大コスト超過）
  const error = adjustedCost > MAX_COST_PER_IMAGE;
  const errorCode = error ? -3 : null;

  // Opus無料判定（元のリクエストサイズで判定）
  const isOpusFree = isOpusFreeGeneration(
    params.width,
    params.height,
    params.steps,
    charRefCount,
    tier,
  );

  // 課金対象枚数
  const billableImages = Math.max(0, nSamples - (isOpusFree ? 1 : 0));
  const generationCost = adjustedCost * billableImages;

  // Vibeコスト（キャラクター参照使用時・Inpaint時は無効）
  let vibeEncodeCost = 0;
  let vibeBatchCost = 0;
  if (charRefCount === 0 && mode !== 'inpaint') {
    vibeEncodeCost = vibeUnencodedCount * VIBE_ENCODE_PRICE;
    vibeBatchCost = calcVibeBatchCost(vibeCount);
  }

  // キャラクター参照コスト
  const charRefCost = charRefCount > 0 ? calcCharRefCost(charRefCount, nSamples) : 0;

  // 合計コスト
  const totalCost = generationCost + charRefCost + vibeEncodeCost + vibeBatchCost;

  return {
    baseCost,
    smeaMultiplier,
    perImageCost,
    strengthMultiplier,
    adjustedCost,
    isOpusFree,
    billableImages,
    generationCost,
    charRefCost,
    vibeEncodeCost,
    vibeBatchCost,
    totalCost,
    error,
    errorCode,
  };
}


// =============================================================================
// Augmentコスト計算
// =============================================================================

/**
 * Augmentツールのコストを計算
 * MAX_PIXELSでクランプ → AUGMENT_MIN_PIXELSまで拡大 → V4基本コスト計算
 * bg-removalは追加の乗数と加算あり
 * @param params Augmentパラメータ
 * @returns コスト計算結果
 */
export function calculateAugmentCost(params: AugmentCostParams): AugmentCostResult {
  const tier = params.tier ?? 0;
  const originalPixels = params.width * params.height;

  // MAX_PIXELSでクランプ
  const clamped = clampToMaxPixels(params.width, params.height, MAX_PIXELS);

  // AUGMENT_MIN_PIXELSまで拡大
  const expanded = expandToMinPixels(clamped.width, clamped.height, AUGMENT_MIN_PIXELS);

  // 基本コスト計算（固定ステップ数）
  const baseCost = calcV4BaseCost(expanded.width, expanded.height, AUGMENT_FIXED_STEPS);

  // bg-removalは特別計算
  let finalCost: number;
  if (params.tool === 'bg-removal') {
    finalCost = BG_REMOVAL_MULTIPLIER * baseCost + BG_REMOVAL_ADDEND;
  } else {
    finalCost = baseCost;
  }

  // Opus無料判定（bg-removal以外、かつ拡大後ピクセル数がOPUS_FREE_PIXELS以下）
  const isOpusFree = params.tool !== 'bg-removal'
    && expanded.width * expanded.height <= OPUS_FREE_PIXELS
    && tier >= OPUS_MIN_TIER;

  const effectiveCost = isOpusFree ? 0 : finalCost;

  return {
    originalPixels,
    adjustedWidth: expanded.width,
    adjustedHeight: expanded.height,
    adjustedPixels: expanded.width * expanded.height,
    baseCost,
    finalCost,
    isOpusFree,
    effectiveCost,
  };
}


// =============================================================================
// アップスケールコスト計算
// =============================================================================

/**
 * アップスケールのコストを計算
 * ピクセル数に基づくテーブル引きで価格を決定
 * @param params アップスケールパラメータ
 * @returns コスト計算結果
 */
export function calculateUpscaleCost(params: UpscaleCostParams): UpscaleCostResult {
  const tier = params.tier ?? 0;
  const pixels = params.width * params.height;

  // Opus無料判定
  if (tier >= OPUS_MIN_TIER && pixels <= UPSCALE_OPUS_FREE_PIXELS) {
    return { pixels, cost: 0, isOpusFree: true, error: false, errorCode: null };
  }

  // コストテーブルから該当価格を検索（降順テーブル、最後にマッチしたものが最小閾値）
  let cost: number | null = null;
  for (const [threshold, price] of UPSCALE_COST_TABLE) {
    if (pixels <= threshold) {
      cost = price;
    }
  }

  return {
    pixels,
    cost,
    isOpusFree: false,
    error: cost === null,
    errorCode: cost === null ? -3 : null,
  };
}

File: rust-api/src/anlas.rs
use crate::constants::*;
use crate::error::NovelAIError;

// =============================================================================
// Type Definitions
// =============================================================================

/// Subscription tier (0=Free, 1=Tablet, 2=Scroll, 3=Opus)
pub type SubscriptionTier = u32;

/// SMEA mode
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SmeaMode {
    Off,
    Smea,
    SmeaDyn,
}

/// Generation mode
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum GenerationMode {
    Txt2Img,
    Img2Img,
    Inpaint,
}

/// Augment tool type
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AugmentToolType {
    Colorize,
    Declutter,
    Emotion,
    Sketch,
    Lineart,
    BgRemoval,
}

/// Generation cost calculation parameters
#[derive(Debug, Clone)]
pub struct GenerationCostParams {
    pub width: u32,
    pub height: u32,
    pub steps: u32,
    pub smea: SmeaMode,
    pub mode: GenerationMode,
    pub strength: f64,
    pub n_samples: u32,
    pub tier: SubscriptionTier,
    pub char_ref_count: u32,
    pub vibe_count: u64,
    pub vibe_unencoded_count: u64,
    pub mask_width: Option<u32>,
    pub mask_height: Option<u32>,
}

impl Default for GenerationCostParams {
    fn default() -> Self {
        Self {
            width: DEFAULT_WIDTH,
            height: DEFAULT_HEIGHT,
            steps: DEFAULT_STEPS,
            smea: SmeaMode::Off,
            mode: GenerationMode::Txt2Img,
            strength: 1.0,
            n_samples: 1,
            tier: 0,
            char_ref_count: 0,
            vibe_count: 0,
            vibe_unencoded_count: 0,
            mask_width: None,
            mask_height: None,
        }
    }
}

/// Generation cost calculation result with breakdown
#[derive(Debug, Clone)]
pub struct GenerationCostResult {
    pub base_cost: u64,
    pub smea_multiplier: f64,
    pub per_image_cost: f64,
    pub strength_multiplier: f64,
    pub adjusted_cost: u64,
    pub is_opus_free: bool,
    pub billable_images: u32,
    pub generation_cost: u64,
    pub char_ref_cost: u64,
    pub vibe_encode_cost: u64,
    pub vibe_batch_cost: u64,
    pub total_cost: u64,
    pub error: bool,
    pub error_code: Option<i32>,
}

/// Augment cost calculation parameters
#[derive(Debug, Clone)]
pub struct AugmentCostParams {
    pub tool: AugmentToolType,
    pub width: u32,
    pub height: u32,
    pub tier: SubscriptionTier,
}

/// Augment cost calculation result
#[derive(Debug, Clone)]
pub struct AugmentCostResult {
    pub original_pixels: u64,
    pub adjusted_width: u64,
    pub adjusted_height: u64,
    pub adjusted_pixels: u64,
    pub base_cost: u64,
    pub final_cost: u64,
    pub is_opus_free: bool,
    pub effective_cost: u64,
}

/// Upscale cost calculation parameters
#[derive(Debug, Clone)]
pub struct UpscaleCostParams {
    pub width: u32,
    pub height: u32,
    pub tier: SubscriptionTier,
}

/// Upscale cost calculation result
#[derive(Debug, Clone)]
pub struct UpscaleCostResult {
    pub pixels: u64,
    pub cost: Option<u64>,
    pub is_opus_free: bool,
    pub error: bool,
    pub error_code: Option<i32>,
}

/// Inpaint size correction result
#[derive(Debug, Clone)]
pub struct InpaintCorrectionResult {
    pub corrected: bool,
    pub width: u64,
    pub height: u64,
}

/// Size adjustment result
#[derive(Debug, Clone)]
pub struct SizeResult {
    pub width: u64,
    pub height: u64,
    pub pixels: u64,
}

// =============================================================================
// Validation
// =============================================================================

fn assert_positive_finite_int(value: u32, name: &str) -> Result<(), NovelAIError> {
    if value == 0 {
        return Err(NovelAIError::Validation(format!(
            "{} must be a positive finite integer, got {}",
            name, value
        )));
    }
    Ok(())
}

fn assert_finite_range(value: f64, min: f64, max: f64, name: &str) -> Result<(), NovelAIError> {
    if !value.is_finite() || value < min || value > max {
        return Err(NovelAIError::Validation(format!(
            "{} must be a finite number between {} and {}, got {}",
            name, min, max, value
        )));
    }
    Ok(())
}

// =============================================================================
// Basic Calculation Functions
// =============================================================================

/// Calculate V4 model base cost.
/// Linear cost calculation based on pixel count and step count.
pub fn calc_v4_base_cost(width: u64, height: u64, steps: u64) -> u64 {
    let pixels = width * height;
    let cost = V4_COST_COEFF_LINEAR * pixels as f64 + V4_COST_COEFF_STEP * pixels as f64 * steps as f64;
    cost.ceil() as u64
}

/// Return the cost multiplier for the given SMEA mode.
pub fn get_smea_multiplier(mode: SmeaMode) -> f64 {
    match mode {
        SmeaMode::SmeaDyn => 1.4,
        SmeaMode::Smea => 1.2,
        SmeaMode::Off => 1.0,
    }
}

/// Check if the generation qualifies for Opus free generation.
pub fn is_opus_free_generation(
    width: u32,
    height: u32,
    steps: u32,
    char_ref_count: u32,
    tier: SubscriptionTier,
    vibe_count: u64,
) -> bool {
    char_ref_count == 0
        && vibe_count == 0
        && (width as u64) * (height as u64) <= OPUS_FREE_PIXELS
        && steps <= OPUS_FREE_MAX_STEPS
        && tier >= OPUS_MIN_TIER
}

/// Calculate Vibe batch cost.
/// Only vibes beyond the free threshold are charged.
pub fn calc_vibe_batch_cost(enabled_vibe_count: u64) -> u64 {
    if enabled_vibe_count > VIBE_FREE_THRESHOLD {
        (enabled_vibe_count - VIBE_FREE_THRESHOLD) * VIBE_BATCH_PRICE
    } else {
        0
    }
}

/// Calculate character reference cost.
pub fn calc_char_ref_cost(char_ref_count: u32, n_samples: u32) -> u64 {
    CHAR_REF_PRICE * char_ref_count as u64 * n_samples as u64
}

// =============================================================================
// Pixel Adjustment Functions
// =============================================================================

/// Expand to minimum pixels (maintaining aspect ratio).
/// No grid snap, Math.floor only.
pub fn expand_to_min_pixels(width: u64, height: u64, min_pixels: u64) -> SizeResult {
    let pixels = width * height;
    if pixels >= min_pixels {
        return SizeResult { width, height, pixels };
    }
    let scale = (min_pixels as f64 / pixels as f64).sqrt();
    let new_w = (width as f64 * scale).ceil() as u64;
    let mut new_h = (height as f64 * scale).floor() as u64;
    // Ensure we actually meet the minPixels requirement
    if new_w * new_h < min_pixels {
        new_h = (height as f64 * scale).ceil() as u64;
    }
    SizeResult {
        width: new_w,
        height: new_h,
        pixels: new_w * new_h,
    }
}

/// Clamp to maximum pixels (maintaining aspect ratio).
/// No grid snap, Math.floor only.
pub fn clamp_to_max_pixels(width: u64, height: u64, max_pixels: u64) -> SizeResult {
    let pixels = width * height;
    if pixels <= max_pixels {
        return SizeResult { width, height, pixels };
    }
    let scale = (max_pixels as f64 / pixels as f64).sqrt();
    let new_w = (width as f64 * scale).floor() as u64;
    let new_h = (height as f64 * scale).floor() as u64;
    SizeResult {
        width: new_w,
        height: new_h,
        pixels: new_w * new_h,
    }
}

// =============================================================================
// Inpaint Correction
// =============================================================================

/// Calculate inpaint mask size correction.
/// If the mask is smaller than the threshold, expand to OPUS_FREE_PIXELS and grid snap.
pub fn calc_inpaint_size_correction(mask_width: u64, mask_height: u64) -> InpaintCorrectionResult {
    if mask_width == 0 || mask_height == 0 {
        return InpaintCorrectionResult {
            corrected: false,
            width: mask_width,
            height: mask_height,
        };
    }

    let pixels = mask_width as f64 * mask_height as f64;
    let threshold = OPUS_FREE_PIXELS as f64 * INPAINT_THRESHOLD_RATIO;

    if pixels >= threshold {
        return InpaintCorrectionResult {
            corrected: false,
            width: mask_width,
            height: mask_height,
        };
    }

    let scale = (OPUS_FREE_PIXELS as f64 / pixels).sqrt();
    let grid = GRID_SIZE as f64;
    let new_w = (((mask_width as f64 * scale).floor() / grid).floor() * grid).max(grid);
    let new_h = (((mask_height as f64 * scale).floor() / grid).floor() * grid).max(grid);

    InpaintCorrectionResult {
        corrected: true,
        width: new_w as u64,
        height: new_h as u64,
    }
}

// =============================================================================
// Main: Generation Cost Calculation
// =============================================================================

/// Calculate generation cost (main orchestrator).
/// Integrates all cost elements and returns the final cost with breakdown.
pub fn calculate_generation_cost(params: &GenerationCostParams) -> Result<GenerationCostResult, NovelAIError> {
    // Input validation
    assert_positive_finite_int(params.width, "width")?;
    assert_positive_finite_int(params.height, "height")?;
    assert_positive_finite_int(params.steps, "steps")?;

    match params.mode {
        GenerationMode::Img2Img | GenerationMode::Inpaint => {
            assert_finite_range(params.strength, 0.0, 1.0, "strength")?;
        }
        _ => {}
    }

    // Validate maskWidth/maskHeight pair
    if params.mode == GenerationMode::Inpaint {
        let has_mask_w = params.mask_width.is_some();
        let has_mask_h = params.mask_height.is_some();
        if has_mask_w != has_mask_h {
            return Err(NovelAIError::Validation(
                "maskWidth and maskHeight must both be specified or both omitted for inpaint mode".to_string(),
            ));
        }
    }

    // Determine effective width/height (inpaint mask correction)
    let mut effective_width = params.width as u64;
    let mut effective_height = params.height as u64;

    if params.mode == GenerationMode::Inpaint {
        if let (Some(mw), Some(mh)) = (params.mask_width, params.mask_height) {
            let correction = calc_inpaint_size_correction(mw as u64, mh as u64);
            if correction.corrected {
                effective_width = correction.width;
                effective_height = correction.height;
            }
        }
    }

    // Base cost calculation
    let base_cost = calc_v4_base_cost(effective_width, effective_height, params.steps as u64);

    // SMEA multiplier
    let smea_multiplier = get_smea_multiplier(params.smea);
    let per_image_cost = base_cost as f64 * smea_multiplier;

    // Strength multiplier (txt2img is always 1.0)
    let strength_multiplier = match params.mode {
        GenerationMode::Txt2Img => 1.0,
        GenerationMode::Img2Img | GenerationMode::Inpaint => params.strength,
    };

    // Adjusted cost (minimum MIN_COST_PER_IMAGE guaranteed)
    let adjusted_cost = std::cmp::max(
        (per_image_cost * strength_multiplier).ceil() as u64,
        MIN_COST_PER_IMAGE,
    );

    // Error check (exceeds max cost)
    if adjusted_cost > MAX_COST_PER_IMAGE {
        return Err(NovelAIError::Validation(format!(
            "Adjusted cost per image ({}) exceeds maximum allowed ({})",
            adjusted_cost, MAX_COST_PER_IMAGE
        )));
    }

    // Opus free check (using corrected dimensions for consistency with cost calculation)
    let is_opus_free = is_opus_free_generation(
        effective_width as u32,
        effective_height as u32,
        params.steps,
        params.char_ref_count,
        params.tier,
        params.vibe_count,
    );

    // Billable images
    let billable_images = if is_opus_free && params.n_samples > 0 {
        params.n_samples - 1
    } else {
        params.n_samples
    };
    let generation_cost = adjusted_cost * billable_images as u64;

    // Vibe cost (disabled when using character reference or inpaint)
    let (vibe_encode_cost, vibe_batch_cost) = if params.char_ref_count == 0 && params.mode != GenerationMode::Inpaint {
        (
            params.vibe_unencoded_count * VIBE_ENCODE_PRICE,
            calc_vibe_batch_cost(params.vibe_count),
        )
    } else {
        (0, 0)
    };

    // Character reference cost
    let char_ref_cost = if params.char_ref_count > 0 {
        calc_char_ref_cost(params.char_ref_count, params.n_samples)
    } else {
        0
    };

    let total_cost = generation_cost + char_ref_cost + vibe_encode_cost + vibe_batch_cost;

    Ok(GenerationCostResult {
        base_cost,
        smea_multiplier,
        per_image_cost,
        strength_multiplier,
        adjusted_cost,
        is_opus_free,
        billable_images,
        generation_cost,
        char_ref_cost,
        vibe_encode_cost,
        vibe_batch_cost,
        total_cost,
        error: false,
        error_code: None,
    })
}

// =============================================================================
// Augment Cost Calculation
// =============================================================================

/// Calculate augment tool cost.
/// Clamp to MAX_PIXELS -> expand to AUGMENT_MIN_PIXELS -> V4 base cost.
/// bg-removal has additional multiplier and addend.
pub fn calculate_augment_cost(params: &AugmentCostParams) -> Result<AugmentCostResult, NovelAIError> {
    assert_positive_finite_int(params.width, "width")?;
    assert_positive_finite_int(params.height, "height")?;

    let tier = params.tier;
    let original_pixels = params.width as u64 * params.height as u64;

    // Clamp to MAX_PIXELS
    let clamped = clamp_to_max_pixels(params.width as u64, params.height as u64, MAX_PIXELS);

    // Expand to AUGMENT_MIN_PIXELS
    let expanded = expand_to_min_pixels(clamped.width, clamped.height, AUGMENT_MIN_PIXELS);

    // Base cost (fixed steps)
    let base_cost = calc_v4_base_cost(expanded.width, expanded.height, AUGMENT_FIXED_STEPS as u64);

    // bg-removal has special calculation
    let final_cost = if params.tool == AugmentToolType::BgRemoval {
        let cost = BG_REMOVAL_MULTIPLIER as f64 * base_cost as f64 + BG_REMOVAL_ADDEND as f64;
        cost.ceil() as u64
    } else {
        base_cost
    };

    // Opus free check (bg-removal is always charged)
    let is_opus_free = params.tool != AugmentToolType::BgRemoval
        && expanded.width * expanded.height <= OPUS_FREE_PIXELS
        && tier >= OPUS_MIN_TIER;

    let effective_cost = if is_opus_free { 0 } else { final_cost };

    Ok(AugmentCostResult {
        original_pixels,
        adjusted_width: expanded.width,
        adjusted_height: expanded.height,
        adjusted_pixels: expanded.width * expanded.height,
        base_cost,
        final_cost,
        is_opus_free,
        effective_cost,
    })
}

// =============================================================================
// Upscale Cost Calculation
// =============================================================================

/// Calculate upscale cost.
/// Table lookup based on pixel count.
pub fn calculate_upscale_cost(params: &UpscaleCostParams) -> Result<UpscaleCostResult, NovelAIError> {
    assert_positive_finite_int(params.width, "width")?;
    assert_positive_finite_int(params.height, "height")?;

    let tier = params.tier;
    let pixels = params.width as u64 * params.height as u64;

    // Opus free check
    if tier >= OPUS_MIN_TIER && pixels <= UPSCALE_OPUS_FREE_PIXELS {
        return Ok(UpscaleCostResult {
            pixels,
            cost: Some(0),
            is_opus_free: true,
            error: false,
            error_code: None,
        });
    }

    // Look up cost in table (ascending, return first match)
    for &(threshold, price) in UPSCALE_COST_TABLE {
        if pixels <= threshold {
            return Ok(UpscaleCostResult {
                pixels,
                cost: Some(price),
                is_opus_free: false,
                error: false,
                error_code: None,
            });
        }
    }

    // No match in table -> error
    Ok(UpscaleCostResult {
        pixels,
        cost: None,
        is_opus_free: false,
        error: true,
        error_code: Some(-3),
    })
}
